### PR TITLE
api: add error flag no_cluster_found

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
As of today `no route found` and route is found but the cluster in that route is not ready are sharing the same error detail.
However, this make big difference. The latter is likely a RDS config problem, or explicit blackhole.
The latter might be a slow CDS.

Signed-off-by: Yuchen Dai <silentdai@gmail.com>

Additional Description:
Address #15448

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by htuch — COMMENTED** [review]
@lambdai API LGTM, I'm trying to spot where this is actually set in the code, can you point this out?

**Comment by lambdai** [comment]
@htuch Thanks!
Not yet set in code. I can either update this this PR or use the next PR

**Comment by lambdai** [comment]
@htuch  Prefer separate PRs: I am updating tcp_proxy. I can also give it a try on HCM. Might need other filter owners to adopt the new flag.

**Comment by lambdai** [comment]
@htuch Both the HCM and the tcp_proxy set NC flag now. usage.rst and VERSION updated as well

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -104,7 +104,7 @@ Removed Config or Runtime
 
 New Features
 ------------
-
+* access log: added a new response flag for upstream cluster not found.
```
Comment (htuch) [inline docs/root/version_history/current.rst:107]: Maybe be a bit more explicit in what this is.

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM, just one tiny doc ask.
/wait

File: docs/root/version_history/current.rst (hunk 1 above)
Comment (lambdai) [reply docs/root/version_history/current.rst]: Added. What else do I need?

**Review by htuch — APPROVED** [review]
LGTM, thanks!

File: include/envoy/stream_info/stream_info.h [hunk 2]
```diff
@@ -82,10 +82,12 @@ enum ResponseFlag {
...
   // Request or connection exceeded the downstream connection duration.
   DurationTimeout = 0x400000,
+  // No cluster found for a given request.
+  NoClusterFound = 0x800000,
   // Upstream response had an HTTP protocol error
-  UpstreamProtocolError = 0x800000,
+  UpstreamProtocolError = 0x1000000,
   // ATTENTION: MAKE SURE THIS REMAINS EQUAL TO THE LAST FLAG.
-  LastFlag = UpstreamProtocolError
+  LastFlag = NoClusterFound
```
Comment (mattklein123) [inline include/envoy/stream_info/stream_info.h:90]: @lambdai this looks like a merge issue. Can you please fix this?

File: docs/root/version_history/current.rst [hunk 3]
```diff
@@ -108,7 +108,7 @@ Removed Config or Runtime
 
 New Features
 ------------
-
+* access log: added the new response flag `NC` for upstream cluster not found. The error flag is set when the http or tcp route is found for the request but the cluster is not available.
```
[the remark this answers is not in the document]
Comment (lambdai) [reply docs/root/version_history/current.rst:111]: https://github.com/envoyproxy/envoy/pull/15538

**Pull request merged**